dir.c

```diff
@@ -1387,25 +1387,37 @@ static enum path_treatment treat_directory(struct dir_struct *dir,
 		return path_none;
 
 	case index_nonexistent:
-		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES)
-			break;
+		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES) {
+			if (!(dir->flags & DIR_HIDE_EMPTY_DIRECTORIES))
+				return exclude ? path_excluded : path_untracked;
+
+			untracked = lookup_untracked(dir->untracked, untracked,
+				     dirname + baselen, len - baselen);
+			return read_directory_recursive(dir, istate, dirname, len,
+					untracked, 1, 0, pathspec);
+		}
```

Comment:
Maybe we can keep the pre-existing code, then (modulo the addition of the `0, ` in the parameter list of `read_directory_recursive()`)? I do not see much benefit of pulling that code into the `switch` statement over how it was done before...